compute_endpoint/globus_compute_endpoint/engines/base.py

```diff
@@ -121,83 +137,92 @@ def set_working_dir(self, run_dir: str | None = None):
+        :param exception_history: List of task exceptions (from previous attempts)
+
+        :returns: packed Result object (bytes)
         """
-        code, user_message = get_result_error_details(exception)
-        error_details = {"code": code, "user_message": user_message}
         execution_end = TaskTransition(
             timestamp=time.time_ns(),
             actor=ActorName.INTERCHANGE,
             state=TaskState.EXEC_END,
         )
-        exception_string = ""
-        for index, prev_exc in enumerate(
-            self._retry_table[task_id]["exception_history"]
-        ):
-            templated_history = _EXC_HISTORY_TMPL.format(
-                ndx=index + 1, exc=get_error_string(exc=prev_exc)
-            )
-            exception_string += templated_history
 
-        final = _EXC_HISTORY_TMPL.format(
-            ndx="final attempt", exc=get_error_string(exc=exception)
-        )
-        exception_string += final
+        *exc_history, last_exc = exception_history
+        code, user_message = get_result_error_details(last_exc)
+        error_details = {"code": code, "user_message": user_message}
+
+        if exc_history:
+            exception_string = "\n".join(
+                _EXC_HISTORY_TMPL.format(ndx=index, exc=get_error_string(exc=exc))
+                for index, exc in itertools.chain(
+                    enumerate(exc_history), (("final attempt", last_exc),)
+                )
+            )
+        else:
+            exception_string = _EXC_NO_HISTORY_TMPL.format(
+                exc=get_error_string(exc=last_exc)
+            )
 
-        result_message = dict(
+        res = Result(
             task_id=task_id,
             data=exception_string,
             exception=exception_string,
             error_details=error_details,
             task_statuses=[execution_begin, execution_end],  # only timings we have
         )
-        return messagepack.pack(Result(**result_message))
-
-    def _setup_future_done_callback(self, task_id: str, future: Future) -> None:
-        """
-        Set up the done() callback for the provided future.
+        return messagepack.pack(res)
 
-        The done callback handles
-        Callback to post result to the passthrough queue
-        Parameters
-        ----------
-        future: Future for which the callback is triggered
-        """
+    def _invoke_submission(
+        self,
+        task_fut: GCEPFuture,
+        submission_partial: t.Callable[..., Future],
+        retry_count: int = 0,
+        exception_history: list | None = None,
+    ):
+        if exception_history is None:
+            exception_history = []
 
         exec_beg = TaskTransition(  # Reminder: used by *closure*, below
             timestamp=time.time_ns(),
             actor=ActorName.INTERCHANGE,
             state=TaskState.WAITING_FOR_LAUNCH,
         )
 
-        def _done_cb(f: Future):
+        def _done_cb(f: GCEPFuture):
```

Comment:
This should be a `concurrent.futures.Future`

Reply:
If I'd listened to Mypy on the _other_ end.  :facepalm:  Thanks!